osf/management/commands/sync_datacite_doi_metadata.py

```diff
@@ -10,31 +10,38 @@
 
 logger = logging.getLogger(__name__)
 
-def sync_datacite_doi_metadata(dry_run=True, retries=4):
+def sync_datacite_doi_metadata(dry_run=True, batch_size=100):
     content_type = ContentType.objects.get_for_model(Registration)
     reg_ids = Identifier.objects.filter(category='doi', content_type=content_type, deleted__isnull=True).values_list(
         'object_id',
         flat=True
     )
 
-    registrations = Registration.objects.exclude(id__in=reg_ids)
-    logger.info(f'{registrations.count()} registrations to mint')
+    registrations = Registration.objects.exclude(
+        deleted__isnull=False,
+        is_public=False,
+        moderation_state='withdrawn'
+    ).exclude(  # Strangely we are required to call exclude two times for a list comp.
+        id__in=reg_ids
+    )[:batch_size]
+    logger.info(f'{"[DRY RUN]: " if dry_run else ""}'
+                f'{registrations.count()} registrations to mint')
+
     for registration in registrations:
-        if not dry_run:
+        retries = 4
+        for i in reversed(range(retries)):
             try:
-                doi = registration.request_identifier('doi')['doi']
-                registration.set_identifier_value('doi', doi)
+                if not dry_run:
+                    doi = registration.request_identifier('doi')['doi']
+                    registration.set_identifier_value('doi', doi)
+                    break
             except DataCiteForbiddenError as e:
-                # Just rate limiting, sleep and retry
-                logger.info(f'retrying for {registration._id}, {retries} retries left')
-                if retries < 1:
+                if i < 1:
                     raise e
```

Comment:
what are the likely causes of a datacite error? should a single error be grounds for interrupting the entire script, or should we log the error and move on to the next registration?

Reply:
The error could really be anything here. I'm catching the `DataCiteForbiddenError` because that's a common error I get when making DOIs too fast, but it's really hard to say what's optimal. I suspect it's a moot point because under normal circumstances it should produce no errors.